File: typescript/tests/payment-service.test.ts
/**
 * Test file for payment service
 * Run with: npm test
 */

import * as db from '../src/database/db';
import { PaymentService } from '../src/payment/service';

describe('Payment Service Tests', () => {
  beforeEach(() => {
    // Reset database before each test
    const connection = db.getConnection();
    Object.keys(connection.transactions).forEach(key => delete connection.transactions[key]);
    Object.keys(connection.users).forEach(key => delete connection.users[key]);
  });

  test('Concurrent Payments', () => {
    // Reset database
    db.getConnection().users['user_test'] = { balance: 1000.0 };
    
    const service = new PaymentService();
    
    // Simulate concurrent payments
    const results: string[] = [];
    const promises: Promise<void>[] = [];
    
    const processPayment = async () => {
      try {
        const paymentId = service.processPayment(
          'user_test', 100.0, 'paypal',
          { email: 'test@example.com' }
        );
        results.push(paymentId);
      } catch (e: any) {
        results.push(`Error: ${e.message}`);
      }
    };
    
    // Create multiple promises to process payments simultaneously
    for (let i = 0; i < 5; i++) {
      promises.push(processPayment());
    }
    
    return Promise.all(promises).then(() => {
      // Check for duplicate payment IDs
      console.log(`Results: ${results}`);
      const uniqueResults = new Set(results);
      console.log(`Unique payments: ${uniqueResults.size}`);
      
      // Check user balance
      const balance = db.getUserBalance('user_test');
      console.log(`Final balance: ${balance}`);
      console.log(`Expected balance: 500.0 (1000 - 5*100)`);
      console.log(`Actual balance: ${balance}`);
    });
  });

  test('Multiple Refunds', () => {
    // Reset database
    db.getConnection().users['user_test2'] = { balance: 1000.0 };
    
    const service = new PaymentService();
    
    // Process a payment
    const paymentId = service.processPayment(
      'user_test2', 100.0, 'paypal',
      { email: 'test@example.com' }
    );
    
    const initialBalance = db.getUserBalance('user_test2');
    console.log(`Balance after payment: ${initialBalance}`);
    
    // Process refund multiple times
    service.processRefund(paymentId, 100.0);
    service.processRefund(paymentId, 100.0);
    service.processRefund(paymentId, 100.0);
    
    const finalBalance = db.getUserBalance('user_test2');
    console.log(`Balance after 3 refunds: ${finalBalance}`);
    console.log(`Expected balance: 1000.0`);
    console.log(`Actual balance: ${finalBalance}`);
  });
});

File: go/tests/payment_service_test.go
package tests

import (
	"fmt"
	"payment-microservice/src/database"
	"payment-microservice/src/payment"
	"sync"
	"testing"
)

// TestConcurrentPayments demonstrates race conditions
func TestConcurrentPayments(t *testing.T) {
	// Reset database
	transactions, users := database.GetConnection()
	users["user_test"] = &database.User{Balance: 1000.0}
	_ = transactions // suppress unused variable
	
	service := &payment.PaymentService()
	
	// Simulate concurrent payments
	var wg sync.WaitGroup
	results := make([]string, 0)
	var mu sync.Mutex
	
	processPayment := func() {
		defer wg.Done()
		paymentID, err := service.ProcessPayment(
			"user_test", 100.0, "paypal",
			map[string]string{"email": "test@example.com"},
		)
		mu.Lock()
		if err != nil {
			results = append(results, fmt.Sprintf("Error: %v", err))
		} else {
			results = append(results, paymentID)
		}
		mu.Unlock()
	}
	
	// Create multiple goroutines to process payments simultaneously
	for i := 0; i < 5; i++ {
		wg.Add(1)
		go processPayment()
	}
	
	wg.Wait()
	
	// Check for duplicate payment IDs (race condition)
	fmt.Printf("Results: %v\n", results)
	uniqueResults := make(map[string]bool)
	for _, r := range results {
		uniqueResults[r] = true
	}
	fmt.Printf("Unique payments: %d\n", len(uniqueResults))
	
	// Check user balance
	balance := database.GetUserBalance("user_test")
	fmt.Printf("Final balance: %.2f\n", balance)
	fmt.Printf("Expected balance: 500.0 (1000 - 5*100)\n")
	fmt.Printf("Actual balance: %.2f\n", balance)
}

// TestDuplicateRefund demonstrates lack of idempotency
func TestDuplicateRefund(t *testing.T) {
	// Reset database
	transactions, users := database.GetConnection()
	users["user_test2"] = &database.User{Balance: 1000.0}
	_ = transactions // suppress unused variable
	
	service := &payment.PaymentService()
	
	// Process a payment
	paymentID, err := service.ProcessPayment(
		"user_test2", 100.0, "paypal",
		map[string]string{"email": "test@example.com"},
	)
	if err != nil {
		t.Fatalf("Failed to process payment: %v", err)
	}
	
	initialBalance := database.GetUserBalance("user_test2")
	fmt.Printf("Balance after payment: %.2f\n", initialBalance)
	
	// Process refund multiple times (should be idempotent but isn't)
	service.ProcessRefund(paymentID, 100.0)
	service.ProcessRefund(paymentID, 100.0)
	service.ProcessRefund(paymentID, 100.0)
	
	finalBalance := database.GetUserBalance("user_test2")
	fmt.Printf("Balance after 3 refunds: %.2f\n", finalBalance)
	fmt.Printf("Expected balance: 1000.0 (should only refund once)\n")
	fmt.Printf("Actual balance: %.2f\n", finalBalance)
}

